sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonFunctions.scala

```diff
@@ -307,3 +308,140 @@ case class GetJsonObject(json: Expression, path: Expression)
+  }
+
+  // and count the number of foldable fields, we'll use this later to optimize evaluation
+  @transient private lazy val constantFields: Int = foldableFieldNames.count(_ != null)
+
+  override lazy val dataType: StructType = {
+    val fields = fieldExpressions.zipWithIndex.map {
+      case (_, idx) => StructField(
+        name = s"c$idx", // mirroring GenericUDTFJSONTuple.initialize
+        dataType = StringType,
+        nullable = true)
+    }
+
+    StructType(fields)
+  }
+
+  override def prettyName: String = "json_tuple"
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.length < 2) {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName requires at least two arguments")
+    } else if (children.forall(child => StringType.acceptsType(child.dataType))) {
+      TypeCheckResult.TypeCheckSuccess
+    } else {
+      TypeCheckResult.TypeCheckFailure(s"$prettyName requires that all arguments are strings")
+    }
+  }
+
+  override def eval(input: InternalRow): InternalRow = {
+    try {
+      val json = jsonExpr.eval(input).asInstanceOf[UTF8String]
+      if (json == null) {
+        return nullRow
+      }
+
+      val parser = jsonFactory.createParser(json.getBytes)
+
+      // only objects are supported
+      if (parser.nextToken() != JsonToken.START_OBJECT) {
+        return nullRow
+      }
+
+      // evaluate the field names as String rather than UTF8String to
+      // optimize lookups from the json token, which is also a String
+      val fieldNames = if (constantFields == fieldExpressions.length) {
+        // typically the user will provide the field names as foldable expressions
+        // so we can use the cached copy
+        foldableFieldNames
+      } else if (constantFields == 0) {
+        // none are foldable so all field names need to be evaluated from the input row
+        fieldExpressions.map(_.eval(input).asInstanceOf[UTF8String].toString)
+      } else {
+        // if there is a mix of constant and non-constant expressions
+        // prefer the cached copy when available
+        foldableFieldNames.zip(fieldExpressions).map {
+          case (null, expr) => expr.eval(input).asInstanceOf[UTF8String].toString
+          case (fieldName, _) => fieldName
+        }
+      }
+
+      val row = Array.ofDim[Any](fieldNames.length)
+
+      // start reading through the token stream, looking for any requested field names
+      while (parser.nextToken() != JsonToken.END_OBJECT) {
+        if (parser.getCurrentToken == JsonToken.FIELD_NAME) {
+          // check to see if this field is desired in the output
+          val idx = fieldNames.indexOf(parser.getCurrentName)
+          if (idx >= 0) {
+            // it is, copy the child tree to the correct location in the output row
+            val output = new ByteArrayOutputStream()
+
+            // write the output directly to UTF8 encoded byte array
+            val generator = jsonFactory.createGenerator(output, JsonEncoding.UTF8)
```

Comment:
I'm going to move the generator construction into the `if` block, it's not used otherwise.